cmd/snapd-apparmor/main_test.go

```diff
@@ -0,0 +1,182 @@
+	// lxc/lxd name should result in a container with internal policy
+	f.Seek(0, 0)
+	f.WriteString("lxc-foo")
+	c.Assert(snapd_apparmor.IsContainerWithInternalPolicy(), Equals, true)
+}
+
+func (s *mainSuite) TestLoadAppArmorProfiles(c *C) {
+	parserCmd := testutil.MockCommand(c, "apparmor_parser", "")
+	defer parserCmd.Restore()
+	err := snapd_apparmor.LoadAppArmorProfiles()
+	c.Assert(err, IsNil)
+	// since no profiles to load the parser should not have been called
+	c.Assert(parserCmd.Calls(), DeepEquals, [][]string(nil))
+
+	// mock a profile
+	err = os.MkdirAll(dirs.SnapAppArmorDir, 0755)
+	c.Assert(err, IsNil)
+
+	profile := filepath.Join(dirs.SnapAppArmorDir, "foo")
+	f, err := os.Create(profile)
+	c.Assert(err, IsNil)
+	f.Close()
+
+	// ensure SNAPD_DEBUG is set in the environment so then --quiet
+	// will *not* be included in the apparmor_parser arguments (since
+	// when these test are run in via CI SNAPD_DEBUG is set)
+	os.Setenv("SNAPD_DEBUG", "1")
+	err = snapd_apparmor.LoadAppArmorProfiles()
+	c.Assert(err, IsNil)
+
+	// check arguments to the parser are as expected
+	c.Assert(parserCmd.Calls(), DeepEquals, [][]string{
+		{"apparmor_parser", "--replace", "--write-cache",
+			"-O", "no-expr-simplify",
+			fmt.Sprintf("--cache-loc=%s/var/cache/apparmor", dirs.GlobalRootDir),
+			profile}})
+
+	// test error case
+	testutil.MockCommand(c, "apparmor_parser", "exit 1")
+	err = snapd_apparmor.LoadAppArmorProfiles()
+	c.Check(err.Error(), Equals, fmt.Sprintf("cannot load apparmor profiles: exit status 1\napparmor_parser output:\n"))
+
+	// rename so file is ignored
+	err = os.Rename(profile, profile+"~")
+	c.Assert(err, IsNil)
+	err = snapd_apparmor.LoadAppArmorProfiles()
+	c.Assert(err, IsNil)
+}
+
+func (s *mainSuite) TestIsContainer(c *C) {
+	c.Check(snapd_apparmor.IsContainer(), Equals, false)
+
+	detectCmd := testutil.MockCommand(c, "systemd-detect-virt", "")
+	defer detectCmd.Restore()
+
+	c.Check(snapd_apparmor.IsContainer(), Equals, true)
+	c.Assert(detectCmd.Calls(), DeepEquals, [][]string{
+		{"systemd-detect-virt", "--quiet", "--container"}})
+}
+
+func (s *mainSuite) TestValidateArgs(c *C) {
+	testCases := []struct {
+		args   []string
+		errMsg string
+	}{
+		{
+			args:   []string{"start"},
+			errMsg: "",
+		},
+		{
+			args:   []string{"foo"},
+			errMsg: "Expected to be called with a single 'start' argument.",
+		},
+		{
+			args:   []string{"start", "foo"},
+			errMsg: "Expected to be called with a single 'start' argument.",
+		},
+	}
+	for _, tc := range testCases {
+		err := snapd_apparmor.ValidateArgs(tc.args)
+		if err != nil {
+			c.Check(err.Error(), Equals, tc.errMsg)
+		} else {
+			c.Check(tc.errMsg, Equals, "")
+		}
+	}
+}
+
+func (s *mainSuite) TestRun(c *C) {
+	os.Args = []string{"snapd-apparmor", "start"}
+	err := snapd_apparmor.Run()
+	c.Assert(err, IsNil)
+
+	// simulate being inside a container environment
+	detectCmd := testutil.MockCommand(c, "systemd-detect-virt", "echo wsl")
```

Comment:
Same comment as above - mocking systemd-detect-virt here it's not really testing much (unless I miss something), I could change this to `echo foo` or remove it and nothing would fail.

Reply:
Yes that is true... more than happy to remove this one if you still want.

Reply:
I think something like https://paste.ubuntu.com/p/STKyNDSvbC/ is what I would propose